### PR TITLE
HHH-18693 Hibernate Processor does not handle inner types

### PR DESCRIPTION
**Pull request by cigaly**
See Jira Issue [HHH-18693](https://hibernate.atlassian.net/browse/HHH-18693)

This is still work in progress, not to be merged with main branch, at least not immediately.

For each (static) inner non-private class one meta-model class will be generated. 

Naming is as follows. If we have:
```
class A {
    ...
    @Entity
    static class B {
        ...
    }
    ...
}
```

Generated class will be named `A_B_`

----------------------
By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt)
and can be relicensed under the terms of the [LGPL v2.1 license](https://www.gnu.org/licenses/old-licenses/lgpl-2.1.txt) in the future at the maintainers' discretion.
For more information on licensing, please check [here](https://github.com/hibernate/hibernate-orm/blob/main/CONTRIBUTING.md#legal).

----------------------

[HHH-18693]: https://hibernate.atlassian.net/browse/HHH-18693?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by cigaly** [comment]
Some existing test classes were changed to make inner classes non-private. Otherwise  such class could not be referenced from meta model class.

**Comment by gavinking** [comment]
Thanks for working on this @cigaly !

**Comment by gavinking** [comment]
> Some existing test classes were changed to make inner classes non-private.  Otherwise such class could not be referenced from meta model class.

I think the Processor is still going to have to be able to elegantly handle `private` inner classes; we don't want to simply _disallow_ that.

So I'm not quit sure what "elegantly" means here ... perhaps it just means "skip them".

**Comment by cigaly** [comment]
> So I'm not quit sure what "elegantly" means here ... perhaps it just means "skip them".

Other than disallowing private classes (either inner or top-level) , excluding them from metamodel generation is only solution that will not cause Java compilation errors.

Now only one test class has to be changed to avoid compilation errors is `org.hibernate.orm.test.schematools.PrimaryKeyColumnOrderTest.EntityId`. Here metemodel generated from

```
	public static class EntityId implements Serializable {
		private int A;
		private int B;
	}
```

will result in

```
public abstract class PrimaryKeyColumnOrderTest_EntityId_ {
	public static final String A = "A";
	public static final String B = "B";
	public static volatile SingularAttribute<EntityId, Integer> A;
	public static volatile SingularAttribute<EntityId, Integer> B;
	public static volatile EmbeddableType<EntityId> class_;
}
```

which is definitely something Java is happy with. Perhaps some mechanism can be added to prevent such duplications, but until then this class should be slightly changed to let tests to run.

**Comment by gavinking** [comment]
> Now only one test class has to be changed to avoid compilation errors is

Hurm.... so that's actually a hole in the algorithm defined by the spec, isn't it? We should fix that, in both Processor, and in the spec. For example, we could say that `A_` -> `_A`.

**Comment by cigaly** [comment]
This is more like one of those "tricky" questions for entertainment pages of some Java magazine. Something like "how are named getters and setters in a class with int property named abc and String property named Abc".

It seems that in current code only problematic are properties with name consisting of single uppercase letter. To fix that (at least, until some other solution) I will change method `org.hibernate.processor.util.StringUtil#getUpperUnderscoreCaseFromLowerCamelCase` by inserting right after method declaration lines

```
		if (lowerCamelCaseString.length()== 1 && lowerCamelCaseString.toUpperCase().equals( lowerCamelCaseString )) {
			return "_" + lowerCamelCaseString;
		}
```

(or, maybe, to use underscore as a suffix?)

If nothing else, this will allow implement metamodel class generation for inner classes, without touching any core (main or test) code.

**Comment by gavinking** [comment]
Yeah but you can just use `Character.isUpperCase()` 😜

**Comment by cigaly** [comment]
Yes, you are right. It seems that I am still sleepy  :-)

**Comment by gavinking** [comment]
@cigaly can you give me permission to push to your branch?

**Comment by gavinking** [comment]
Or, whatever, don't worry, I can open a new PR myself, I guess.

**Comment by cigaly** [comment]
> @cigaly can you give me permission to push to your branch?

Of course ... only I hope that I did not done something completely different :-)

**Comment by gavinking** [comment]
> Or, whatever, don't worry, I can open a new PR myself, I guess.

Heh, I actually typed these comments on the wrong PR. I was meaning to post on the MySQL PR.

File: tooling/metamodel-generator/src/main/java/org/hibernate/processor/HibernateProcessor.java [hunk 1]
```diff
@@ -362,59 +365,70 @@ private void processClasses(RoundEnvironment roundEnvironment) {
...
+	}
+
+	private void processElement(Element element) {
+		try {
+			if ( !included( element )
+					|| hasAnnotation( element, Constants.EXCLUDE )
+					|| hasPackageAnnotation( element, Constants.EXCLUDE ) ) {
+				// skip it completely
+			}
+			else if ( isEntityOrEmbeddable( element ) && !element.getModifiers().contains( Modifier.PRIVATE )) {
```
Comment (gavinking) [inline tooling/metamodel-generator/src/main/java/org/hibernate/processor/HibernateProcessor.java:379]: I think we could just say that the processor completely skips `private` inner types. WDYT of this @cigaly:

```suggestion
			if ( !included( element )
					|| hasAnnotation( element, Constants.EXCLUDE )
					|| hasPackageAnnotation( element, Constants.EXCLUDE ) 
					|| element.getModifiers().contains( Modifier.PRIVATE ) ) {
				// skip it completely
			}
			else if ( isEntityOrEmbeddable( element ) ) {
```

File: tooling/metamodel-generator/src/main/java/org/hibernate/processor/HibernateProcessor.java [hunk 2]
```diff
@@ -714,7 +728,7 @@ private void writeIndex() {
 					.createResource(
 							StandardLocation.SOURCE_OUTPUT,
 							ENTITY_INDEX,
-							entityName,
+							URLEncoder.encode(entityName, UTF_8),
```
Comment (gavinking) [inline tooling/metamodel-generator/src/main/java/org/hibernate/processor/HibernateProcessor.java:731]: Why is this necessary?

File: tooling/metamodel-generator/src/main/java/org/hibernate/processor/util/StringUtil.java [hunk 3]
```diff
@@ -98,6 +98,9 @@ public static String nameToMethodName(String name) {
 	}
 
 	public static String getUpperUnderscoreCaseFromLowerCamelCase(String lowerCamelCaseString) {
+		if ( lowerCamelCaseString.length() == 1 && Character.isUpperCase( lowerCamelCaseString.charAt( 0 ) ) ) {
```
Comment (gavinking) [inline tooling/metamodel-generator/src/main/java/org/hibernate/processor/util/StringUtil.java:101]:
```suggestion
		if ( lowerCamelCaseString.length() == 1 && isUpperCase( lowerCamelCaseString.charAt( 0 ) ) ) {
```

**Review by gavinking — COMMENTED** [review]
Thanks, @cigaly, looks pretty reasonable, but see my comments.

File: tooling/metamodel-generator/src/main/java/org/hibernate/processor/HibernateProcessor.java (hunk 1 above)
Comment (cigaly) [reply tooling/metamodel-generator/src/main/java/org/hibernate/processor/HibernateProcessor.java:379]: Agree. And for maybe will be good idea to keep access modifier for non private classes?

Comment (gavinking) [reply tooling/metamodel-generator/src/main/java/org/hibernate/processor/HibernateProcessor.java:379]: > And for maybe will be good idea to keep access modifier for non private classes?

You mean generate the metamodel with the same visibility? Yeah, that makes sense I suppose...

File: tooling/metamodel-generator/src/main/java/org/hibernate/processor/HibernateProcessor.java (hunk 2 above)
Comment (cigaly) [reply tooling/metamodel-generator/src/main/java/org/hibernate/processor/HibernateProcessor.java:731]: Take look at the `Person` in `org.hibernate.orm.test.hql.QuotedIdentifierTest`. There entity name is `The Person` (guess that this is not forbidden by the specification) and processing (and/or compilation) is crashing with message:

`java.lang.IllegalArgumentException: Invalid relative name: The Person`

If you like, I can write similar test for Hibernate Processor. Of course, if such entity name is allowed.

File: tooling/metamodel-generator/src/main/java/org/hibernate/processor/HibernateProcessor.java (hunk 1 above)
Comment (cigaly) [reply tooling/metamodel-generator/src/main/java/org/hibernate/processor/HibernateProcessor.java]: One more thing ... in PR [#9220](https://github.com/hibernate/hibernate-orm/pull/9220) id class is generated  as inner class of parent meta model class. Here meta model class for each inner class is generated as separate class. I guess that now is good time to decide if this should stay that way or will be better idea to generate them as inner classes as well.

File: tooling/metamodel-generator/src/main/java/org/hibernate/processor/HibernateProcessor.java (hunk 2 above)
Comment (gavinking) [reply tooling/metamodel-generator/src/main/java/org/hibernate/processor/HibernateProcessor.java:732]: OK, I see. But then I think we need to make the same change in `ProcessorSessionFactory.findEntityByUnqualifiedName()`?

>  if such entity name is allowed.

Well that is a _great_ question. It looks like the JPA spec is silent on the format of entity names, except to say that they are case-sensitive, and unique in a PU.

I think, however, that we can infer that they are not supposed to have spaces in them, because then we would not be able to parse the query language.

I think I should open an issue against the spec.

File: tooling/metamodel-generator/src/main/java/org/hibernate/processor/HibernateProcessor.java (hunk 1 above)
Comment (gavinking) [reply tooling/metamodel-generator/src/main/java/org/hibernate/processor/HibernateProcessor.java]: I think I would prefer it to be an inner class.

File: tooling/metamodel-generator/src/main/java/org/hibernate/processor/HibernateProcessor.java (hunk 2 above)
Comment (cigaly) [reply tooling/metamodel-generator/src/main/java/org/hibernate/processor/HibernateProcessor.java:732]: No, `ProcessorSessionFactory.findEntityByUnqualifiedName()` will work fine, since entity name is key for `entityNameMappings`. And such entity can be, despite its really 'ugly' name, used in HQL queries if properly quoted. `QuotedIdentifierTest` is testing exactly that. I've just tried to name class 'The Person?', and it is working fine. 
But, I guess that spec should be changes anyway. Although, I guess that some hardcore Star Trek fans will be disappointed if not able to name entites using Klingon alphabet :-) [It is working now, BTW]

File: tooling/metamodel-generator/src/main/java/org/hibernate/processor/HibernateProcessor.java (hunk 1 above)
Comment (cigaly) [reply tooling/metamodel-generator/src/main/java/org/hibernate/processor/HibernateProcessor.java]: Agree, I will try to refactor this to make things that way some time next week.

File: tooling/metamodel-generator/src/main/java/org/hibernate/processor/validation/ProcessorSessionFactory.java [hunk 4]
```diff
@@ -495,7 +495,7 @@ boolean isAttributeDefined(String entityName, String fieldName) {
 			&& findPropertyByPath(entityClass, fieldName, getDefaultAccessType(entityClass)) != null;
 	}
 
-	private TypeElement findEntityClass(String entityName) {
+	public TypeElement findEntityClass(String entityName) {
```
Comment (gavinking) [inline tooling/metamodel-generator/src/main/java/org/hibernate/processor/validation/ProcessorSessionFactory.java:498]:
```suggestion
	@Override
	public TypeElement findEntityClass(String entityName) {
```

Comment (cigaly) [reply tooling/metamodel-generator/src/main/java/org/hibernate/processor/validation/ProcessorSessionFactory.java:499]: :+1:

**Comment by cigaly** [comment]
Code changes to generate metadata for inner class inside other metadata class instead of new metadata class file for each managed class.

Note: This code will surely need some (beauty and other) changes.

File: tooling/metamodel-generator/src/main/java/org/hibernate/processor/HibernateProcessor.java (hunk 2 above)
Comment (cigaly) [reply tooling/metamodel-generator/src/main/java/org/hibernate/processor/HibernateProcessor.java]: I've removed that change and excluded all "troublemakers". Unfortunately, I can not find wither where this entity index is specified or used, Problem with such "ugly" entity names can not be solved in Hibernate, because finally it is used in  `com.sun.tools.javac.file.JavacFileManager#getFileForOutput`. That name must be encoded to pass as constructor parameter for `java.net.URI`, but must not be encoded to allow file to be created with proper entity name. Something that Joseph Keller called "Catch 22" :-)

Comment (gavinking) [reply tooling/metamodel-generator/src/main/java/org/hibernate/processor/HibernateProcessor.java]: I don't think we need to worry about indexing such pathological entity names. They can't be used in queries anyway (since the query can't be parsed).

Comment (gavinking) [reply tooling/metamodel-generator/src/main/java/org/hibernate/processor/HibernateProcessor.java]: https://github.com/jakartaee/persistence/issues/686

For the record.

Comment (cigaly) [reply tooling/metamodel-generator/src/main/java/org/hibernate/processor/HibernateProcessor.java]: Can be used if quoted. But I agree that that such names should not be used. Maybe only to put some disclaimer about that for those who like to sue company for mot specifically saying that cat can not be put in microwave oven to dry? :-)

Comment (cigaly) [reply tooling/metamodel-generator/src/main/java/org/hibernate/processor/HibernateProcessor.java]: It seems that Hibernate will happily accept non-ASCII letters in entity (and/or class) names (PostgreSQL even in column names), so only it should be not big "sacrifice" to not allow special characters

Comment (gavinking) [reply tooling/metamodel-generator/src/main/java/org/hibernate/processor/HibernateProcessor.java]: > Can be used if quoted. 

Ahhh uuhmmm, yeah, OK sure. I forgot HQL has quoted identifiers (JPQL doesn't).

File: tooling/metamodel-generator/src/main/java/org/hibernate/processor/annotation/NonManagedMetamodel.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+		final NonManagedMetamodel metamodel =
+				new NonManagedMetamodel( element, context, jakartaDataStaticMetamodel, parent );
+		if ( parent != null ) {
+			metamodel.setParentElement( parent.getElement() );
+			parent.addInnerClass( metamodel );
+		}
+		return metamodel;
+	}
+
+	protected void init() {
```
Comment (gavinking) [inline tooling/metamodel-generator/src/main/java/org/hibernate/processor/annotation/NonManagedMetamodel.java:31]: @cigaly Can you add the missing `@Override` please?

**Review by gavinking — APPROVED** [review]
OK, I've tried this out and it looks like it's working great!

@cigaly could you do me the favor of squashing it down to two or three commits so that I can merge it?

Thanks!!!

File: tooling/metamodel-generator/src/main/java/org/hibernate/processor/annotation/NonManagedMetamodel.java (hunk 5 above)
Comment (cigaly) [reply tooling/metamodel-generator/src/main/java/org/hibernate/processor/annotation/NonManagedMetamodel.java:31]: Done! Local tests are passing, so I guess that I did not messed anything :-)

**Comment by gavinking** [comment]
Thanks so much @cigaly, this is great!
